{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Copyright **`(c)`** 2022 Giovanni Squillero `<squillero@polito.it>`  \n",
    "[`https://github.com/squillero/computational-intelligence`](https://github.com/squillero/computational-intelligence)  \n",
    "Free for personal or classroom use; see [`LICENSE.md`](https://github.com/squillero/computational-intelligence/blob/master/LICENSE.md) for details.  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "source": [
    "# Lab 3: ES\n",
    "\n",
    "## Task\n",
    "\n",
    "Write agents able to play [*Nim*](https://en.wikipedia.org/wiki/Nim), with an arbitrary number of rows and an upper bound $k$ on the number of objects that can be removed in a turn (a.k.a., *subtraction game*).\n",
    "\n",
    "The goal of the game is to **avoid** taking the last object.\n",
    "\n",
    "* Task2.1: An agent using fixed rules based on *nim-sum* (i.e., an *expert system*)\n",
    "* Task2.2: An agent using evolved rules using ES\n",
    "\n",
    "## Instructions\n",
    "\n",
    "* Create the directory `lab2` inside your personal course repository for the course \n",
    "* Put a `README.md` and your solution (all the files, code and auxiliary data if needed)\n",
    "\n",
    "## Notes\n",
    "\n",
    "* Working in group is not only allowed, but recommended (see: [Ubuntu](https://en.wikipedia.org/wiki/Ubuntu_philosophy) and [Cooperative Learning](https://files.eric.ed.gov/fulltext/EJ1096789.pdf)). Collaborations must be explicitly declared in the `README.md`.\n",
    "* [Yanking](https://www.emacswiki.org/emacs/KillingAndYanking) from the internet is allowed, but sources must be explicitly declared in the `README.md`.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "from pprint import pprint, pformat\n",
    "from collections import namedtuple\n",
    "from random import random, choice, randint, seed, gauss\n",
    "from copy import deepcopy, copy\n",
    "import numpy as np\n",
    "import networkx as nx\n",
    "import matplotlib.pyplot as plt\n",
    "from functools import reduce"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The *Nim* and *Nimply* classes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "Nimply = namedtuple(\"Nimply\", \"row, num_objects\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Nim:\n",
    "    def __init__(self, num_rows: int, k: int = None) -> None:\n",
    "        self._rows = [i * 2 + 1 for i in range(num_rows)]\n",
    "        self._k = k\n",
    "\n",
    "    def __bool__(self):\n",
    "        return sum(self._rows) > 0\n",
    "\n",
    "    def __str__(self):\n",
    "        return \"<\" + \" \".join(str(_) for _ in self._rows) + \">\"\n",
    "\n",
    "    @property\n",
    "    def rows(self) -> tuple:\n",
    "        return tuple(self._rows)\n",
    "\n",
    "    def nimming(self, ply: Nimply) -> None:\n",
    "        row, num_objects = ply\n",
    "        assert self._rows[row] >= num_objects\n",
    "        assert self._k is None or num_objects <= self._k\n",
    "        self._rows[row] -= num_objects\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "NUM_PROB = 0.3\n",
    "MAX_FIT = 50000\n",
    "TOURNAMENT_SIZE = 2\n",
    "NUM_POPULATIONS = 12\n",
    "MUTATION_PROBABILITY = .9\n",
    "DIM_CHROM = 10\n",
    "FITNESS_TRESHOLD = 868\n",
    "GENERATIONS = 20\n",
    "OPERATIONS = [\"+\",\"-\",\"*\",\"&\",\"|\",\"^\"]\n",
    "\n",
    "from random import choice, randint, shuffle\n",
    "\n",
    "from random import choice, randint, random\n",
    "\n",
    "#### Generates a random operation gene with some constraints.\n",
    "def generate_operation(index, op_indexes: list, elements_to_add: list) -> tuple:\n",
    "    op = (choice(OPERATIONS),)\n",
    "    \n",
    "    if elements_to_add:         #if not all piles are already in an operation, add them\n",
    "        if op_indexes:  \n",
    "            if random() < 0.5:\n",
    "                num1 = choice(op_indexes)\n",
    "                num2 = choice(elements_to_add)\n",
    "                elements_to_add.remove(num2)\n",
    "            else:\n",
    "                num2 = choice(op_indexes)\n",
    "                num1 = choice(elements_to_add)\n",
    "                elements_to_add.remove(num1)\n",
    "        else:\n",
    "            # Handle the case when op_indexes is empty\n",
    "            num1 = choice(elements_to_add)\n",
    "            num2 = choice(elements_to_add)\n",
    "            while num2 == num1:  # Ensure num2 is different from num1\n",
    "                num2 = randint(0, index - 1)\n",
    "            elements_to_add.remove(num1)\n",
    "            elements_to_add.remove(num2)\n",
    "    else :\n",
    "        if op_indexes:  # concatenate with previous operations in chromosome\n",
    "            if random() < 0.5:\n",
    "                num1 = choice(op_indexes)\n",
    "                num2 = randint(0,index-1)\n",
    "            else:\n",
    "                num2 = choice(op_indexes)\n",
    "                num1 = randint(0,index-1)\n",
    "        else:\n",
    "            # Handle the case when op_indexes is empty\n",
    "            num1 = randint(0,index-1)\n",
    "            num2 = randint(0,index-1)\n",
    "            while num2 == num1:  # Ensure num2 is different from num1\n",
    "                num2 = randint(0, index - 1)\n",
    "    op += (num1, num2)\n",
    "    op_indexes.append(index)\n",
    "    return op\n",
    "\n",
    "### Randomly changes the operator of the given operation\n",
    "def mutate_operation(operation) -> tuple:\n",
    "    new_op = choice(OPERATIONS)\n",
    "    while new_op == operation[0]:\n",
    "        new_op = choice(OPERATIONS)\n",
    "    new_tuple = (new_op, operation[1], operation[2])\n",
    "    return new_tuple\n",
    "\n",
    "### Checks if a position is safe\n",
    "def check_rules(father_NP: bool, children: list) -> int:\n",
    "    flag = 0\n",
    "    if father_NP:  # No P after P             \n",
    "        for i in range(len(children)):\n",
    "            if children[i]:\n",
    "                flag = 1\n",
    "                break\n",
    "    else:  # At least a P after N\n",
    "        flag = 1\n",
    "        for i in range(len(children)):\n",
    "            if children[i]:\n",
    "                flag = 0\n",
    "                break\n",
    "    return flag\n",
    "\n",
    "### Determines the state of a position (N or P)\n",
    "def setNP(node: tuple, chromosome) -> int:\n",
    "    return chromosome.get_expression(node)\n",
    "    \n",
    "    \n",
    "class Chromosome:\n",
    "    def __init__(self, expression) -> None:\n",
    "        self.expr = []\n",
    "        self.expr = list(expression)\n",
    "        self.fitness = 0\n",
    "        self.operations_indexes = []\n",
    "    \n",
    "    ### Generates a chromosome with random expression\n",
    "    def create_random(self, nim_rows):\n",
    "        indexes = list(range(len(nim_rows)))    \n",
    "        shuffle(indexes)\n",
    "        self.expr.extend(indexes)\n",
    "\n",
    "        elements_to_add = [i for i in range(len(nim_rows))]\n",
    "            \n",
    "\n",
    "        for i in range(len(nim_rows), DIM_CHROM-1):\n",
    "            self.expr.append(generate_operation(i, self.operations_indexes, elements_to_add))\n",
    "\n",
    "    ### Decodes the gene\n",
    "    def recursive_decoder(self, index, position_tuple):\n",
    "        if isinstance(self.expr[index], int):\n",
    "            return position_tuple[self.expr[index]]\n",
    "        else:\n",
    "            operator = self.expr[index][0]\n",
    "            if operator == \"+\":\n",
    "                op1 = self.recursive_decoder(self.expr[index][1], position_tuple)\n",
    "                op2 = self.recursive_decoder(self.expr[index][2], position_tuple)\n",
    "                result = op1 + op2\n",
    "            elif operator == \"*\":\n",
    "                op1 = self.recursive_decoder(self.expr[index][1], position_tuple)\n",
    "                op2 = self.recursive_decoder(self.expr[index][2], position_tuple)\n",
    "                result = op1 * op2\n",
    "            elif operator == \"-\":\n",
    "                op1 = self.recursive_decoder(self.expr[index][1], position_tuple)\n",
    "                op2 = self.recursive_decoder(self.expr[index][2], position_tuple)\n",
    "                result = abs(op1 - op2)\n",
    "            elif operator == \"&\":\n",
    "                op1 = self.recursive_decoder(self.expr[index][1], position_tuple)\n",
    "                op2 = self.recursive_decoder(self.expr[index][2], position_tuple)\n",
    "                result = op1 & op2\n",
    "            elif operator == \"|\":\n",
    "                op1 = self.recursive_decoder(self.expr[index][1], position_tuple)\n",
    "                op2 = self.recursive_decoder(self.expr[index][2], position_tuple)\n",
    "                result = op1 | op2\n",
    "            elif operator == \"^\":\n",
    "                op1 = self.recursive_decoder(self.expr[index][1], position_tuple)\n",
    "                op2 = self.recursive_decoder(self.expr[index][2], position_tuple)\n",
    "                result = op1 ^ op2\n",
    "            else:\n",
    "                # Handle unknown operator\n",
    "                raise ValueError(f\"Unknown operator: {operator}\")\n",
    "\n",
    "            #print(f\"Operator: {operator}, op1: {op1}, op2: {op2}, result: {result}\")\n",
    "            return result\n",
    "\n",
    "    ### Returns the value of the expression for a certain board\n",
    "    def get_expression(self, position_tuple) -> int:\n",
    "        return self.recursive_decoder(len(self.expr) - 1, position_tuple)\n",
    "\n",
    "    ### Print expression encoded in the gene\n",
    "    def print_expression(self, index, position_tuple):\n",
    "        if isinstance(self.expr[index], int):\n",
    "            print(f\"{position_tuple[self.expr[index]]}\", end=\"\")\n",
    "            return position_tuple[self.expr[index]]\n",
    "        else:\n",
    "            operator = self.expr[index][0]\n",
    "            print(\"(\", end=\"\")\n",
    "            op1 = self.print_expression(self.expr[index][1], position_tuple)\n",
    "            print(f\" {operator} \", end=\"\")\n",
    "            op2 = self.print_expression(self.expr[index][2], position_tuple)\n",
    "            print(\")\", end=\"\")\n",
    "            return f\"({op1} {operator} {op2})\"\n",
    "\n",
    "    ### Computes fitness of the chromosome\n",
    "    def fitness_function(self, state: Nim) -> int:\n",
    "        def node_fitness(node: tuple, visited: dict) -> int:\n",
    "            # If sub-tree fitness is already known, just return it\n",
    "            if node in visited:\n",
    "                return visited[node][0]\n",
    "\n",
    "            # Terminal case, fitness 0 if P, else 1\n",
    "            if all(value == 0 for value in node):\n",
    "                result = setNP(node, self)\n",
    "                curr_state = True if result == 0 else False\n",
    "                if curr_state == True:\n",
    "                    visited[node] = (0, curr_state)\n",
    "                    return 0\n",
    "                else:\n",
    "                    visited[node] = (0, curr_state)\n",
    "                    return 0\n",
    "\n",
    "            fit = 0\n",
    "            children_state = []\n",
    "\n",
    "            # Determine state of the current node (P or N)\n",
    "            result = setNP(node, self)\n",
    "            curr_state = True if result == 0 else False\n",
    "\n",
    "            # Compute or retrieve fitness for all children sub-trees\n",
    "            for i, value in enumerate(node):\n",
    "                for new_value in range(value, -1, -1): \n",
    "                    if new_value != value:\n",
    "                        child = node[:i] + (new_value,) + node[i + 1:]\n",
    "                        child_fit = node_fitness(child, visited)\n",
    "                        children_state.append(visited[child][1]) \n",
    "                        fit += child_fit\n",
    "            \n",
    "            # Now add current fitness\n",
    "            fit += check_rules(curr_state, children_state)\n",
    "            # Add node to visited\n",
    "            visited[node] = (fit, curr_state)\n",
    "\n",
    "            return fit\n",
    "        \n",
    "        visited = {}\n",
    "        self.fitness = node_fitness(state.rows, visited) \n",
    "        return self.fitness\n",
    "\n",
    "    ### Changes a random gene in the chromosome\n",
    "    def mutate(self, state: Nim) -> 'Chromosome':\n",
    "        offspring = deepcopy(self)\n",
    "        pos = randint(0, len(self.expr) - 1)\n",
    "\n",
    "        if isinstance(self.expr[pos], int):\n",
    "            # If it's an integer, change it to a random row index\n",
    "            offspring.expr[pos] = randint(0, len(state.rows) - 1)\n",
    "        else:\n",
    "            # If it's an operation, change it to a new operation\n",
    "            offspring.expr[pos] = mutate_operation(self.expr[pos])\n",
    "\n",
    "        offspring.fitness = None  \n",
    "\n",
    "        return offspring\n",
    "\n",
    "\n",
    "### Exchanges genes of two chromosomes\n",
    "def one_cut_xover(c1: Chromosome, c2: Chromosome) -> list:\n",
    "    exchanged = []\n",
    "    cut_point = randint(0, len(c1.expr) - 1)\n",
    "    #print(f\"cut point is {cut_point}\")\n",
    "    off_1 = c1.expr[:cut_point] + c2.expr[cut_point:]\n",
    "    off_2 = c2.expr[:cut_point] + c1.expr[cut_point:]\n",
    "\n",
    "    coff_1 = Chromosome(off_1)\n",
    "    coff_2 = Chromosome(off_2)\n",
    "    exchanged.append(coff_1)\n",
    "    exchanged.append(coff_2)\n",
    "    return exchanged\n",
    "\n",
    "\n",
    "### Evolves dna one step\n",
    "def run(dna: list, state: Nim) -> list[Chromosome]: \n",
    "    for k in range(0, len(dna), 4):  # Iterate over dna 4 by 4\n",
    "        p1 = randint(0, 3)  # Select two parents\n",
    "        p2 = randint(0, 3)\n",
    "        while p1 == p2:\n",
    "            p1 = randint(0, 3)\n",
    "        off = generate_offspring(dna[k + p1], dna[k + p2], state)  # Generate one offspring\n",
    "        dna.append(off)        \n",
    "\n",
    "    sorted_dna = sorted(dna, key=lambda x: x.fitness)       # sort ascending and remove last N/4 elements\n",
    "    result_dna = sorted_dna[:-int(NUM_POPULATIONS/4)] if int(NUM_POPULATIONS/4) > 0 else sorted_dna[:]\n",
    "    shuffle(result_dna)\n",
    "\n",
    "    return result_dna\n",
    "\n",
    "### Generates a chromosome starting from two parents    \n",
    "def generate_offspring(parent1 : Chromosome, parent2 : Chromosome, state: Nim) -> Chromosome:\n",
    "    \n",
    "    off = one_cut_xover(parent1,parent2)    #crossover\n",
    "    for i in range(2):\n",
    "        if(random() < MUTATION_PROBABILITY):    #offspring mutation\n",
    "            off[i] = off[i].mutate(state)\n",
    "    fit1 = off[0].fitness_function(state)\n",
    "    fit2 = off[1].fitness_function(state)\n",
    "    if fit1>=fit2:\n",
    "        return off[0]    \n",
    "    else:\n",
    "        return off[1]\n",
    "    \n",
    "### Evolves dna\n",
    "def evolve(dna: list, state: Nim) -> list:\n",
    "    if max(dna, key=lambda x: x.fitness).fitness == 0:      #dna not yet initialized\n",
    "        for ch in dna:\n",
    "            ch.fitness = ch.fitness_function(state)\n",
    "\n",
    "    for _ in range(GENERATIONS):\n",
    "        dna = run(dna, state)\n",
    "\n",
    "    dna = sorted(dna, key = lambda x: x.fitness)\n",
    "    \n",
    "    return dna\n",
    "\n",
    "### Prints dna\n",
    "def print_dna(dna):\n",
    "    for ch in dna:\n",
    "        print(f\"{ch.fitness}\")   \n",
    "    print() \n",
    "\n",
    "### Creates a number of random chromosomes\n",
    "def create_population(state: Nim) -> list:\n",
    "    dna = []\n",
    "    for _ in range(NUM_POPULATIONS):\n",
    "        cromo = Chromosome([])\n",
    "        cromo.create_random(state.rows)\n",
    "        cromo.fitness_function(state)\n",
    "        dna.append(cromo)\n",
    "    return dna\n",
    "\n",
    "### Creates a number of bad chromosomes (fitness higher than a threshold)\n",
    "def dummy_population(state : Nim) -> list:\n",
    "    dna = []\n",
    "    i = 0\n",
    "    while i<NUM_POPULATIONS:\n",
    "        cromo = Chromosome([])\n",
    "        cromo.create_random(state.rows)\n",
    "        if cromo.fitness_function(state) > FITNESS_TRESHOLD:\n",
    "            dna.append(cromo)\n",
    "            i+=1\n",
    "    return dna\n",
    "\n",
    "### Prints Nim board\n",
    "def print_nim_board(board):\n",
    "    max_height = max(board)\n",
    "\n",
    "    for height in range(max_height, 0, -1):\n",
    "        row = ['*' if col >= height else ' ' for col in board]\n",
    "        print(' '.join(row))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sample (and silly) startegies "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def pure_random(state: Nim, nothing) -> Nimply:\n",
    "    \"\"\"A completely random move\"\"\"\n",
    "    row = choice([r for r, c in enumerate(state.rows) if c > 0])\n",
    "    num_objects = randint(1, state.rows[row])\n",
    "    return Nimply(row, num_objects)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gabriele(state: Nim, nothing) -> Nimply:\n",
    "    \"\"\"Pick always the maximum possible number of the lowest row\"\"\"\n",
    "    possible_moves = [(r, o) for r, c in enumerate(state.rows) for o in range(1, c + 1)]\n",
    "    return Nimply(*max(possible_moves, key=lambda m: (-m[0], m[1])))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "def adaptive(state: Nim, nothing) -> Nimply:\n",
    "    \"\"\"A strategy that can adapt its parameters\"\"\"\n",
    "    genome = {\"love_small\": 0.5}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "\n",
    "def nim_sum(state: Nim) -> int:\n",
    "    tmp = np.array([tuple(int(x) for x in f\"{c:032b}\") for c in state.rows])\n",
    "    xor = tmp.sum(axis=0) % 2\n",
    "    return int(\"\".join(str(_) for _ in xor), base=2)\n",
    "\n",
    "def analize(raw: Nim) -> dict:\n",
    "    cooked = dict()\n",
    "    cooked[\"possible_moves\"] = dict()\n",
    "    for ply in (Nimply(r, o) for r, c in enumerate(raw.rows) for o in range(1, c + 1)):\n",
    "        tmp = deepcopy(raw)\n",
    "        tmp.nimming(ply)\n",
    "        cooked[\"possible_moves\"][ply] = nim_sum(tmp)\n",
    "    return cooked\n",
    "\n",
    "def optimal(state: Nim, nothing) -> Nimply:\n",
    "    analysis = analize(state)\n",
    "    logging.debug(f\"analysis:\\n{pformat(analysis)}\")\n",
    "    spicy_moves = [ply for ply, ns in analysis[\"possible_moves\"].items() if ns == 0]\n",
    "    if not spicy_moves:\n",
    "        spicy_moves = list(analysis[\"possible_moves\"].keys())\n",
    "    ply = choice(spicy_moves)\n",
    "    return ply\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Evolutionary Strategy\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "def evo_nim_sum(state:Nim, ch: Chromosome ) ->int:\n",
    "    return ch.get_expression(state.rows)\n",
    "\n",
    "def evo_analize(raw: Nim, ch: Chromosome) -> dict:\n",
    "    cooked = dict()\n",
    "    cooked[\"possible_moves\"] = dict()\n",
    "    for ply in (Nimply(r, o) for r, c in enumerate(raw.rows) for o in range(1, c + 1)):\n",
    "        tmp = deepcopy(raw)\n",
    "        tmp.nimming(ply)\n",
    "        cooked[\"possible_moves\"][ply] = evo_nim_sum(tmp, ch)\n",
    "    return cooked\n",
    "\n",
    "def evolutionary(state :Nim, ch: Chromosome) -> Nimply:\n",
    "    analysis = evo_analize(state, ch)\n",
    "    spicy_moves =[ply for ply, ns in analysis[\"possible_moves\"].items() if ns == 0] \n",
    "    if not spicy_moves:\n",
    "        spicy_moves = list(analysis[\"possible_moves\"].keys())\n",
    "    ply = choice(spicy_moves)\n",
    "    return ply\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Match Simulator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:init : <1 3 5>\n",
      "INFO:root:ply: player 0 plays Nimply(row=2, num_objects=2)\n",
      "INFO:root:status: (1, 3, 3)\n",
      "INFO:root:ply: player 1 plays Nimply(row=0, num_objects=1)\n",
      "INFO:root:status: (0, 3, 3)\n",
      "INFO:root:ply: player 0 plays Nimply(row=1, num_objects=3)\n",
      "INFO:root:status: (0, 0, 3)\n",
      "INFO:root:ply: player 1 plays Nimply(row=2, num_objects=3)\n",
      "INFO:root:status: (0, 0, 0)\n",
      "INFO:root:status: Player 1 won!\n"
     ]
    }
   ],
   "source": [
    "logging.getLogger().setLevel(logging.INFO)\n",
    "args = []\n",
    "\n",
    "##################################\n",
    "# SET YOUR STRATEGIES HERE #######\n",
    "#   pure_random\n",
    "#   gabriele\n",
    "#   adaptive\n",
    "#   optimal\n",
    "#   evolutionary\n",
    "##################################\n",
    "strategy = (evolutionary, optimal)\n",
    "\n",
    "\n",
    "\n",
    "nim = Nim(3)\n",
    "winners = []\n",
    "\n",
    "\n",
    "if strategy[0] == evolutionary:\n",
    "    dna1 = create_population(nim)\n",
    "    dna1 = evolve(dna1, nim)\n",
    "    ns1 = dna1[0]\n",
    "    args.append(ns1)\n",
    "    #print(\"player 0 using expr\")\n",
    "   # ns1.print_expression(len(ns1.expr)-1, nim.rows)\n",
    "   # print()\n",
    "    #print(f\"with fitness {ns1.fitness}\")\n",
    "\n",
    "if strategy[1] == evolutionary:\n",
    "    dna2 = create_population(nim)\n",
    "    dna2 = evolve(dna2, nim)\n",
    "    ns2 = dna2[0]\n",
    "    args.append(ns2)\n",
    "    #print(\"player 1 using expr\")\n",
    "    #ns2.print_expression(len(ns2.expr)-1, nim.rows)\n",
    "    #print()\n",
    "    #print(f\"with fitness {ns2.fitness}\")\n",
    "\n",
    "logging.info(f\"init : {nim}\")\n",
    "player = 0\n",
    "\n",
    "while nim:\n",
    "    if player == 0:\n",
    "        ply = strategy[player](nim, *args)\n",
    "        logging.info(f\"ply: player {player} plays {ply}\")\n",
    "        nim.nimming(ply)\n",
    "        logging.info(f\"status: {nim.rows}\")\n",
    "        player = 1 - player\n",
    "    else:\n",
    "        ply = strategy[player](nim, *args)\n",
    "        logging.info(f\"ply: player {player} plays {ply}\")\n",
    "        nim.nimming(ply)\n",
    "        logging.info(f\"status: {nim.rows}\")\n",
    "        player = 1 - player\n",
    "logging.info(f\"status: Player {1- player} won!\")\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.7 ('ci22-dPIXJ0_o-py3.10')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  },
  "vscode": {
   "interpreter": {
    "hash": "10197e8e2f2aa67e2c349105091c77f4cd384fce4877865f002d9ec653f96bc0"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
